File: java-samples/referencestates-sanctionsbody/workflows/src/main/java/com/r3/developers/samples/referencestate/workflows/GetIOUFlow.java
package com.r3.developers.samples.referencestate.workflows;

import com.r3.developers.samples.referencestate.states.SanctionableIOUState;
import net.corda.v5.application.flows.ClientRequestBody;
import net.corda.v5.application.flows.ClientStartableFlow;
import net.corda.v5.application.flows.CordaInject;
import net.corda.v5.application.marshalling.JsonMarshallingService;
import net.corda.v5.base.annotations.Suspendable;
import net.corda.v5.base.types.MemberX500Name;
import net.corda.v5.ledger.utxo.UtxoLedgerService;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class GetIOUFlow implements ClientStartableFlow {
    @CordaInject
    public JsonMarshallingService jsonMarshallingService;

    @CordaInject
    public UtxoLedgerService ledgerService;

    @NotNull
    @Override
    @Suspendable
    public String call(@NotNull ClientRequestBody requestBody) {

        List<IOU> iouList =
                ledgerService.findUnconsumedStatesByType(SanctionableIOUState.class).stream().map(
                        it -> new IOU(
                                it.getState().getContractState().getValue(),
                                it.getState().getContractState().getLender().getName(),
                                it.getState().getContractState().getBorrower().getName(),
                                it.getState().getContractState().getUniqueIdentifier()
                        )
                ).collect(Collectors.toList());
        return jsonMarshallingService.format(iouList);
    }

    class IOU{
        private int value;
        private MemberX500Name lender;
        private MemberX500Name borrower;
        private UUID uniqueIdentifier;

        public IOU(int value, MemberX500Name lender, MemberX500Name borrower, UUID uniqueIdentifier) {
            this.value = value;
            this.lender = lender;
            this.borrower = borrower;
            this.uniqueIdentifier = uniqueIdentifier;
        }

        public int getValue() {
            return value;
        }

        public void setValue(int value) {
            this.value = value;
        }

        public MemberX500Name getLender() {
            return lender;
        }

        public void setLender(MemberX500Name lender) {
            this.lender = lender;
        }

        public MemberX500Name getBorrower() {
            return borrower;
        }

        public void setBorrower(MemberX500Name borrower) {
            this.borrower = borrower;
        }

        public UUID getUniqueIdentifier() {
            return uniqueIdentifier;
        }

        public void setUniqueIdentifier(UUID uniqueIdentifier) {
            this.uniqueIdentifier = uniqueIdentifier;
        }
    }
}

/* Example JSON to put into REST-API POST requestBody
 {
    "clientRequestId": "get-iou",
    "flowClassName": "com.r3.developers.samples.referencestate.workflows.GetIOUFlow",
    "requestBody": {
    }
 }
 */

File: kotlin-samples/referencestates-sanctionsbody/workflows/src/main/kotlin/com/r3/developers/samples/referencestate/workflows/GetIOUFlow.kt
package com.r3.developers.samples.referencestate.workflows

import com.r3.developers.samples.referencestate.states.SanctionableIOUState
import net.corda.v5.application.flows.ClientRequestBody
import net.corda.v5.application.flows.ClientStartableFlow
import net.corda.v5.application.flows.CordaInject
import net.corda.v5.application.marshalling.JsonMarshallingService
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.base.types.MemberX500Name
import net.corda.v5.ledger.utxo.UtxoLedgerService
import org.slf4j.LoggerFactory
import java.util.*

data class IOU(val value: Int,
               val lender: MemberX500Name,
               val borrower:MemberX500Name,
               val uniqueIdentifier: UUID)


class GetIOUFlow : ClientStartableFlow {

    private companion object {
        val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
    }
    @CordaInject
    lateinit var jsonMarshallingService: JsonMarshallingService

    // Injects the UtxoLedgerService to enable the flow to make use of the Ledger API.
    @CordaInject
    lateinit var ledgerService: UtxoLedgerService

    @Suspendable
    override fun call(requestBody: ClientRequestBody): String {
        log.info("GetIOUFlow.call() called")

        // Queries the VNode's vault for unconsumed states and converts the result to a serializable DTO.
        val states = ledgerService.findUnconsumedStatesByType(SanctionableIOUState::class.java)
        val results = states.map {it ->
            IOU(
                it.state.contractState.value,
                it.state.contractState.lender.name,
                it.state.contractState.borrower.name,
                it.state.contractState.uniqueIdentifier
            )
        }

        // Uses the JsonMarshallingService's format() function to serialize the DTO to Json.
        return jsonMarshallingService.format(results)
    }
}